pkg/envoy/lds/inmesh.go

```diff
@@ -140,6 +151,84 @@ func (lb *listenerBuilder) getInboundMeshHTTPFilterChain(proxyService service.Me
 	return filterChain, nil
 }
 
+func (lb *listenerBuilder) getInboundMeshTCPFilterChain(proxyService service.MeshService, servicePort uint32) (*xds_listener.FilterChain, error) {
+	// Construct TCP filters
+	filters, err := lb.getInboundTCPFilters(proxyService)
+	if err != nil {
+		log.Error().Err(err).Msgf("Error constructing inbound TCP filters for proxy service %s", proxyService)
+		return nil, err
+	}
+
+	// Construct downstream TLS context
+	marshalledDownstreamTLSContext, err := ptypes.MarshalAny(envoy.GetDownstreamTLSContext(proxyService, true /* mTLS */))
+	if err != nil {
+		log.Error().Err(err).Msgf("Error marshalling DownstreamTLSContext for proxy service %s", proxyService)
+		return nil, err
+	}
+
+	filterchainName := fmt.Sprintf("%s:%d", inboundMeshTCPFilterChain, servicePort)
+	return &xds_listener.FilterChain{
+		Name: filterchainName,
+		FilterChainMatch: &xds_listener.FilterChainMatch{
+			// The DestinationPort is the service port the downstream directs traffic to
+			DestinationPort: &wrapperspb.UInt32Value{
+				Value: servicePort,
+			},
+
+			// The ServerName is the SNI set by the downstream in the UptreamTlsContext by GetUpstreamTLSContext()
+			// This is not a field obtained from the mTLS Certificate.
+			ServerNames: []string{proxyService.ServerName()},
+
+			// Only match when transport protocol is TLS
+			TransportProtocol: envoy.TransportProtocolTLS,
+
+			// In-mesh proxies will advertise this, set in the UpstreamTlsContext by GetUpstreamTLSContext()
+			ApplicationProtocols: envoy.ALPNInMesh,
+		},
+		Filters: filters,
+		TransportSocket: &xds_core.TransportSocket{
+			Name: wellknown.TransportSocketTls,
+			ConfigType: &xds_core.TransportSocket_TypedConfig{
+				TypedConfig: marshalledDownstreamTLSContext,
+			},
+		},
+	}, nil
+}
+
+func (lb *listenerBuilder) getInboundTCPFilters(proxyService service.MeshService) ([]*xds_listener.Filter, error) {
+	var filters []*xds_listener.Filter
+
+	// Apply an RBAC filter when permissive mode is disabled. The RBAC filter must be the first filter in the list of filters.
+	if !lb.cfg.IsPermissiveTrafficPolicyMode() {
+		// Apply RBAC policies on the inbound filters based on configured policies
+		rbacFilter, err := lb.buildRBACFilter()
+		if err != nil {
+			log.Error().Err(err).Msgf("Error applying RBAC filter for proxy service %s", proxyService)
+			return nil, err
+		}
+		// RBAC filter should be the very first filter in the filter chain
+		filters = append(filters, rbacFilter)
+	}
+
+	// Apply the TCP Proxy Filter
+	tcpProxy := &xds_tcp_proxy.TcpProxy{
+		StatPrefix:       "inbound-mesh-tcp-proxy",
```

Comment:
I think you will want to use `proxyService` in some form of prefix, to help identify the statistics for this filter later on

Reply:
discussed offline, we'll explore the effect of this going in the future.

Reply:
Discussed with @eduser25 offline and we have decided to look into the usage of stats prefixes in general: https://github.com/openservicemesh/osm/issues/2156